File: 4330backupfiles/PLCbonus/WAE.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
from WAEParser import parser
import sys


def eval(tree):
    exp = ''
    if tree[0] == 'num':
        exp = 'WAENode(nodeType="' + tree[0] + '",numValue=' \
            + str(tree[1]) + ')\n'
    elif tree[0] == 'id':
        exp = 'WAENode(nodeType="' + tree[0] + '",idValue="' \
            + str(tree[1]) + '")'
    elif tree[0] == '+' or tree[0] == '-' or tree[0] == '*' or tree[0] \
        == '/':
        t1 = eval(tree[1])
        t2 = eval(tree[2])
        exp = 'WAENode(nodeType="' + tree[0] + '",child1=' + str(t1) \
            + ',child2=' + str(t2) + ')'
    elif tree[0] == 'with':
        t_arr = []
        for val in tree[1]:
            t_arr.append('("' + str(val[0]) + '",' + eval(val[1]) + ')')
        t1 = ','.join(t_arr)
        t2 = eval(tree[2])
        exp = 'WAENode(nodeType="' + tree[0] + '",vars=List(' + str(t1) \
            + '),child1=' + str(t2) + ')'
    elif tree[0] == 'if':
        t1 = eval(tree[1])
        t2 = eval(tree[2])
        t3 = eval(tree[3])
        exp = 'WAENode(nodeType="' + tree[0] + '",child1=' + str(t1) \
            + ',child2=' + str(t2) + ',child3=' + str(t3) + ')'
    return exp


def main():
    filename = sys.argv[1]
    with open(filename) as fp:
        waes = fp.readlines()
    file = open(r"Driver.scala", 'a')
    lines = ['object Driver extends App { \n', '''import WAE._

''',
             'val d = Map[String,Double]() \n']
    for l in lines:
        file.write(l)
    for (idx, wae) in enumerate(waes):
        try:
            tree = parser.parse(wae)
            if tree is not None:
                exp = eval(tree)
                file.write('val e' + str(idx) + ' = ' + exp)
                file.write('\n')
                file.write('println(eval(e' + str(idx) + ',d))')
                file.write("\n \n")
            else:
                print("else")
        except Exception as abcd:
            print(abcd)
            break

    file.write('} \n')
    file.close()


main()

File: wae.scala
import fiddls.Fiddle,
Fiddls.println
import scalajs.js

@js.annotation.JSExportTopLevel("ScalaFiddle")
object WAE {

  abstract class Node

  case class WAENode(
    nodeType: String,
    numValue: Double = 0.0,
    idValue: String = "",
    child1: WAENode = null,
    child2: WAENode = null,
    child3: WAENode = null,
    vars: List[(String,WAENode)] = List[(String,WAENode)]()
  ) extends Node

  def eval(e: WAENode, d: Map[String,Double]): Double = {
    if (e.nodeType == "num") e.numValue
    else if(e.nodeType == "id") {

      if(d.exists(x => x._1 == e.idValue)) d(e.idValue)
      else throw new Exception("Uninstantiated variable error")
    }
    else if(e.nodeType == "+")
    {
      eval(e.child1,d)+eval(e.child2,d)
    }
    else if(e.nodeType == "-")
    {
      eval(e.child1,d)-eval(e.child2,d)
    }
    else if(e.nodeType == "*")
    {
      eval(e.child1,d)*eval(e.child2,d)
    }
    else if(e.nodeType == "/")
    {
      if(eval(e.child2,d)!=0) eval(e.child1,d)/eval(e.child2,d)
      else throw new Exception("Divide by zero error")
    }
    else if(e.nodeType == "if")
    {
      if (eval(e.child1,d)!=0) eval(e.child2,d)
      else eval(e.child3,d)
    }
    else if(e.nodeType == "with")
    {
      //populate dictionary
      var a = d
      for(l<-e.vars)
        {
          a += (l._1 -> eval(l._2, d))
        }
      //println(a)
      eval(e.child1, a)
    }

    else throw new Exception("Illegal Syntax")
  }

}
